openedx/features/content_tagging/rest_api/v1/views.py

```diff
@@ -0,0 +1,184 @@
+        * 403 - Permission denied
+
+    **Update Parameters**
+        * id (required): - The id of the taxonomy to update
+
+    **Update Request Body**
+        * name (optional): User-facing label used when applying tags from this taxonomy to Open edX objects.
+        * description (optional): Provides extra information for the user when applying tags from this taxonomy
+        to an object.
+        * enabled (optional): Only enabled taxonomies will be shown to authors.
+        * required (optional): Indicates that one or more tags from this taxonomy must be added to an object.
+        * allow_multiple (optional): Indicates that multiple tags from this taxonomy may be added to an object.
+        * allow_free_text (optional): Indicates that tags in this taxonomy need not be predefined; authors may
+        enter their own tag values.
+
+    **Update Example Requests**
+        PUT api/tagging/v1/taxonomy/:id                                            - Update a taxonomy
+        {
+            "name": "Taxonomy New Name",
+            "description": "This is a new description",
+            "enabled": False,
+            "required": False,
+            "allow_multiple": False,
+            "allow_free_text": True,
+        }
+        PATCH api/tagging/v1/taxonomy/:id                                          - Partially update a taxonomy
+        {
+            "name": "Taxonomy New Name",
+        }
+
+    **Update Query Returns**
+        * 200 - Success
+        * 403 - Permission denied
+
+    **Delete Parameters**
+        * id (required): - The id of the taxonomy to delete
+
+    **Delete Example Requests**
+        DELETE api/tagging/v1/taxonomy/:id                                         - Delete a taxonomy
+
+    **Delete Query Returns**
+        * 200 - Success
+        * 404 - Taxonomy not found
+        * 403 - Permission denied
+
+    """
+
+    class TaxonomyPagination(DefaultPagination):
+        page_size = 100
+        max_page_size = 1000
+
+    pagination_class = TaxonomyPagination
+
+    class UserOrgFilterBackend(BaseFilterBackend):
+        """
+        Filter taxonomies based on the user's organization.
+        """
+
+        def filter_queryset(self, request, queryset, view):
+            if is_global_taxonomy_admin(request.user):
+                return queryset
+            else:
+                user_orgs = get_user_taxonomy_orgs(request.user)
+                user_orgs_short_name = [org.short_name for org in user_orgs]
+                return queryset.filter(
+                    Q(
+                        Exists(
+                            TaxonomyOrg.objects.filter(
+                                taxonomy=OuterRef("pk"),
+                                rel_type=TaxonomyOrg.RelType.OWNER,
+                                org__short_name__in=user_orgs_short_name,
+                            )
+                        )
+                    )
+                    | Q(enabled=True)
```

Comment:
> * This filter misses where the user is not a global taxonomy admin, but the TaxonomyOrg has `org=None` (instance taxonomy).

The `enabled=True` handle these cases (because we don't want to show `enabled=False, org=None` to org. But this code changed because we don't want a user from 'orgA' to list taxonomies from 'orgB'

```python
   def filter_queryset(self, request, queryset, view):
            if is_global_taxonomy_admin(request.user):
                return queryset
            else:
                user_orgs = get_user_taxonomy_orgs(request.user)
                user_orgs_short_name = [org.short_name for org in user_orgs]
                return queryset.filter(
                    (
                        Q(enabled=True)
                        & Q(
                            Exists(
                                TaxonomyOrg.objects.filter(
                                    taxonomy=OuterRef("pk"),
                                    rel_type=TaxonomyOrg.RelType.OWNER,
                                    org=None,
                                )
                            )
                        )
                    )
                    | Q(
                        Exists(
                            TaxonomyOrg.objects.filter(
                                taxonomy=OuterRef("pk"),
                                rel_type=TaxonomyOrg.RelType.OWNER,
                                org__short_name__in=user_orgs_short_name,
                            )
                        )
                    )
                )
``

Reply:
> I don't know whether this detailed model filtering logic should live in `content_tagging.models` or `content_tagging.api`, but it shouldn't live here..

I'm not sure if I agree with you on that.

Currently, the API doesn't handle the permissions or "know" the rules. The rules are secured on the ViewSet via `TaxonomyObjectPermissions`. Similarly, I think the 'permission filters' should be part of the view (via  `UserOrgFilterBackend`). 

I also thought about using the `UserOrgFilterBackend` but leaving the filter structure (the `Q` query) in the model. Still, that way, we will inject permission code in the model (and I don't think this is good either).

If you think both (rules and filters) should be the API's responsibility, I'm also good with it. We could call the has_perm inside the API. This will lead to more code because the Django View has a lot of mixins to handle that automagically.

If you are worried about bloating the View code, we could move this class to a `filters.py` (but let the view handle the permission). Then, we have a pair of `rules.py` and `filters.py` for what a user can do and what he is allowed to view.

Do you think this makes sense? Are there other concerns that I'm not getting here?

Reply:
@rpenido I'd really hoped that all the fancy model filtering logic could live in TaxonomyOrg, since that's what understands its internal relationships.. but see what shakes out with [this change](https://github.com/openedx/edx-platform/pull/32871#discussion_r1281250100).

> If you are worried about bloating the View code (and I think it is now), I will move this class to a filters.py (but let the view handle the permission). Then, we have a pair of rules.py and filters.py for what a user can do and what he is allowed to view.

👍